core/src/sumeragi/network_topology.rs

```diff
@@ -5,341 +5,174 @@
+            Some(index) if index == self.min_votes_for_commit() => Role::ProxyTail,
+            Some(_) => Role::ObservingPeer,
+            None => {
+                trace!(%peer_id, "Peer is not in topology.");
+                Role::ObservingPeer
+            }
+        }
     }
-
-    /// Maximum number of faulty peers that the network will tolerate.
-    #[allow(clippy::integer_division)]
-    pub fn max_faults(&self) -> usize {
-        (self.sorted_peers.len() - 1) / 3
+    /// Get leader's peer id.
+    pub fn leader(&self) -> &PeerId {
+        &self.sorted_peers[0]
     }
-
-    /// Create network topology from the list of [`WorldStateView`] peers at the given block
-    pub fn recreate(&mut self, wsv: &WorldStateView, block: HashOf<VersionedCommittedBlock>) {
-        let wsv_peers = wsv.peers_ids().iter().map(|id| id.clone());
-
-        *self = Topology::builder()
-            .with_peers(wsv_peers)
-            .at_block(block)
-            .build(0)
-            .expect("Invalid topology");
+    /// Get proxy tail's peer id.
+    pub fn proxy_tail(&self) -> &PeerId {
+        &self.sorted_peers[self.min_votes_for_commit()]
+    }
+    /// Add or remove peers from the topology.
+    pub fn update_peer_list(&mut self, new_peer_list: &[PeerId]) {
+        // The complexity here is to not two peer_ids that are different but have the same key.
+        let mut new_list: Vec<PeerId> = self
+            .sorted_peers
+            .iter()
+            .filter(|pid| new_peer_list.iter().any(|p| &p == pid))
+            .cloned()
+            .collect();
+        new_list.extend_from_slice(
+            &new_peer_list
+                .iter()
+                .filter(|pid| !self.sorted_peers.iter().any(|p| &p == pid))
+                .cloned()
+                .collect::<Vec<PeerId>>(),
+        );
+        self.sorted_peers = new_list;
+    }
+    /// Rotate peers after each failed attempt to create a block.
+    pub fn rotate_all(&mut self) {
+        let top = self.sorted_peers.remove(0);
+        self.sorted_peers.push(top);
+    }
+    /// Re-arrange the set of peers after each successful block commit.
+    pub fn rotate_set_a(&mut self) {
+        let top = self.sorted_peers.remove(0);
+        self.sorted_peers.insert(
+            self.min_votes_for_commit().min(self.sorted_peers.len()),
+            top,
+        );
+    }
+    /// Pull peers up in the topology to the top of the a set while preserving local order.
+    pub fn lift_up_peers(&mut self, to_lift_up: &[PublicKey]) {
+        let mut observing = Vec::new();
+        let mut i = 0;
+        while i < self.sorted_peers.len() {
+            if to_lift_up.contains(&self.sorted_peers[i].public_key) {
+                i += 1;
+            } else {
+                observing.insert(0, self.sorted_peers.remove(i)); // This has to be insert(0) and not push in order to preserve order.
+            }
+        }
+        self.sorted_peers.extend(observing);
     }
```

Comment:
should I mention that worst case complexity is quite high for this simple operation here?

Reply:
No because N is small and it's all in cache so it's not a problem.